tests/PHPStan/Analyser/nsrt/loose-comparisons.php

```diff
@@ -653,4 +653,85 @@ public function sayInt(
 
 	}
 
+	/**
+	 * @param true|1|"1" $looseOne
+	 * @param false|0|"0" $looseZero
+	 * @param false|1 $constMix
+	 */
+	public function sayConstUnion(
+		$looseOne,
+		$looseZero,
+		$constMix
+	): void
+	{
+		assertType('true', $looseOne == 1);
+		assertType('false', $looseOne == 0);
+		assertType('true', $looseOne == true);
+		assertType('false', $looseOne == false);
+		assertType('true', $looseOne == "1");
+		assertType('false', $looseOne == "0");
+		assertType('false', $looseOne == []);
+
+		assertType('false', $looseZero == 1);
+		assertType('true', $looseZero == 0);
+		assertType('false', $looseZero == true);
+		assertType('true', $looseZero == false);
+		assertType('false', $looseZero == "1");
+		assertType('true', $looseZero == "0");
+		assertType('bool', $looseZero == []);
+
+		assertType('bool', $constMix == 0);
+		assertType('bool', $constMix == 1);
+		assertType('bool', $constMix == true);
+		assertType('bool', $constMix == false);
+		assertType('bool', $constMix == "1");
+		assertType('bool', $constMix == "0");
+		assertType('bool', $constMix == []);
+
+		assertType('true', $looseOne == $looseOne);
+		assertType('true', $looseZero == $looseZero);
+		assertType('false', $looseOne == $looseZero);
+		assertType('false', $looseZero == $looseOne);
+		assertType('bool', $looseOne == $constMix);
+		assertType('bool', $constMix == $looseOne);
+		assertType('bool', $looseZero == $constMix);
+		assertType('bool', $constMix == $looseZero);
+	}
+
+	/**
+	 * @param uppercase-string $upper
+	 * @param lowercase-string $lower
+	 */
+	public function sayIntersection(
+		string $upper,
+		string $lower,
+		string $s,
+	): void
+	{
+		assertType('bool', '' == $upper);
+		if ($upper != '') {
+			assertType('false', '' == $upper);
+		}
+		assertType('bool', '0' == $upper);
+		assertType('bool', 'a' == $upper); // should be false
+		assertType('bool', 'abc' == $upper); // should be false
+		assertType('bool', 'aBc' == $upper);
+		assertType('bool', strtoupper($s) == $upper);
+		assertType('bool', strtolower($s) == $upper); // should be false
+		assertType('bool', $upper == $lower); // should be false
```

Comment:
@VincentLanglet are you interessted in implementing `looseCompare()` in upper/lower-case string type? :)

I tried but can't get my head arround this atm

Reply:
The tests are not true, I edited with comment 
```
		assertType('bool', 'aBc' == $upper); // => Should be false no ?
		
		assertType('bool', strtolower($s) == $upper); // should be false => NO, it can be true with `$s = '&!<42' or an empty string
		
		assertType('bool', $upper == $lower); // should be false => NO, it can be true with `$upper = ''` or `$upper = '&!<42'`
		
		assertType('bool', 'AbC' == $lower); // => Should be false no ?
		
		assertType('bool', strtoupper($s) == $lower); // should be false => NO, it can be true with `$s = '&!<42' or an empty string
		
		assertType('bool', $lower == $upper); // should be false => NO, it can be true with `$lower = ''` or `$lower = '&!<42'`
```

So maybe it's the reason of your difficulties ?

For the implementation I would go with
https://github.com/phpstan/phpstan-src/commit/ff999299810056a8cb419f818a0bb85e0697b758

I had to comment the `if` condition because there is a bug, you might know how to fix
```
// Here PHPStan consider $lower is `lowercase`
if ($lower != '') {
     assertType('false', '' == $lower);
     // Here PHPStan consider $lower is `lowercase&non-empty-string`
}
// Here PHPStan consider $lower is `string`
```

Reply:
See https://phpstan.org/r/5b9fe2de-f827-4e3b-99ac-9d8b62dca596 @staabm

Reply:
thanks for looking into it. I have added your proposal as a separate commit.